xds/internal/resolver/xds_resolver_test.go

```diff
@@ -1106,54 +1110,137 @@ func (s) TestResolverWRR(t *testing.T) {
 	}
 }
 
-func (s) TestXDSResolverMaxStreamDuration(t *testing.T) {
-	xdsR, xdsC, tcc, cancel := testSetup(t, setupOpts{target: target})
-	defer xdsR.Close()
-	defer cancel()
+// TestResolverMaxStreamDuration tests the case where the resolver receives max
+// stream duration as part of the listener and route configuration resources.
+// The test verifies that the RPC timeout returned by the config selector
+// matches the value specified in the config.
+func (s) TestResolverMaxStreamDuration(t *testing.T) {
+	mgmtServer, err := e2e.StartManagementServer(e2e.ManagementServerOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer mgmtServer.Stop()
 
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	waitForWatchListener(ctx, t, xdsC, targetStr)
-	xdsC.InvokeWatchListenerCallback(xdsresource.ListenerUpdate{RouteConfigName: routeStr, MaxStreamDuration: time.Second, HTTPFilters: routerFilterList}, nil)
-	waitForWatchRouteConfig(ctx, t, xdsC, routeStr)
+	// Create a bootstrap configuration specifying the above management server.
+	nodeID := uuid.New().String()
+	cleanup, err := xdsbootstrap.CreateFile(xdsbootstrap.Options{
+		NodeID:    nodeID,
+		ServerURI: mgmtServer.Address,
+		Version:   xdsbootstrap.TransportV3,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
 
-	defer func(oldNewWRR func() wrr.WRR) { newWRR = oldNewWRR }(newWRR)
-	newWRR = testutils.NewTestWRR
+	const serviceName = "my-service-client-side-xds"
+	tcc, rClose := buildResolverForTarget(t, resolver.Target{URL: *testutils.MustParseURL("xds:///" + serviceName)})
+	defer rClose()
 
-	// Invoke the watchAPI callback with a good service update and wait for the
-	// UpdateState method to be called on the ClientConn.
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes: []*xdsresource.Route{{
-					Prefix:            newStringP("/foo"),
-					WeightedClusters:  map[string]xdsresource.WeightedCluster{"A": {Weight: 1}},
-					MaxStreamDuration: newDurationP(5 * time.Second),
-				}, {
-					Prefix:            newStringP("/bar"),
-					WeightedClusters:  map[string]xdsresource.WeightedCluster{"B": {Weight: 1}},
-					MaxStreamDuration: newDurationP(0),
-				}, {
-					Prefix:           newStringP(""),
-					WeightedClusters: map[string]xdsresource.WeightedCluster{"C": {Weight: 1}},
-				}},
+	// Configure the management server with a listener resource that specifies a
+	// max stream duration as part of its HTTP connection manager. Also
+	// configure a route configuration resource, which has multiple routes with
+	// different values of max stream duration.
```

Comment:
Can you also explain the logic of taking these multiple values and mapping to the rpc timeout? I.e. longest of all? How do the values in LDS and RDS interact?

Reply:
I didn't think longest of all was correct, just an example comment of explanation. Can you add this explanation that you just wrote to this comment string (can be as verbose or shorter)?